### PR TITLE
Backend/bug fix/signup fix/#1837

### LINKED ISSUE #1837: "Invalid token" when trying to sign up with username already signed up
Sentry Issue: [BACKEND-2T](https://sentry.io/organizations/couchers/issues/2582011114/?referrer=github_integration)

```
UniqueViolation: duplicate key value violates unique constraint "uq_signup_flows_username"
DETAIL:  Key (username)=(esra) already exists.

  File "sqlalchemy/engine/base.py", line 1771, in _execute_context
    self.dialect.do_execute(
  File "sqlalchemy/engine/default.py", line 717, in do_execute
    cursor.execute(statement, parameters)

IntegrityError: (psycopg2.errors.UniqueViolation) duplicate key value violates unique constraint "uq_signup_flows_username"
DETAIL:  Key (username)=(esra) already exists.

[SQL: UPDATE signup_flows SET username=%(username)s, birthdate=%(birthdate)s, gender=%(gender)s, hosting_status=%(hosting_status)s, city=%(city)s, geom=CAST(CAST(ST_GeogFromText(%(param_1)s) AS geography(POINT,4326)) AS geometry(POINT,4326)), geom_radius=%(geom_radius)s, accepted_tos=%(accepted_tos)s WHERE signup_flows.id = %(signup_flows_id)s]
[param...
(15 additional frame(s) were not displayed)
...
  File "sqlalchemy/engine/base.py", line 1814, in _execute_context
    self._handle_dbapi_exception(
  File "sqlalchemy/engine/base.py", line 1995, in _handle_dbapi_exception
    util.raise_(
  File "sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "sqlalchemy/engine/base.py", line 1771, in _execute_context
    self.dialect.do_execute(
  File "sqlalchemy/engine/default.py", line 717, in do_execute
    cursor.execute(statement, parameters)
```

### PR DESCRIPTION
**Pull request by satvik6005**
"
added signup SIGNUP_FLOW_USERNAME_TAKEN in error.py file and added condition to check unique username during sign up in auth.py 
"
**Backend checklist**
- [x] Formatted my code by running `autoflake -r -i --remove-all-unused-imports src && isort . && black .` in `app/backend`
- [ ] Added tests for any new code or added a regression test if fixing a bug
- [x] All tests pass
- [x] Run the backend locally and it works
- [ ] Added migrations if there are any database changes, rebased onto `develop` if necessary for linear migration history

Closes #1837

### CODE REVIEW

**Comment by lucaslcode** [comment]
Nice debugging work! Someone will review this properly but it would be great to also add a regression test.

**Comment by satvik6005** [comment]
> Nice debugging work! Someone will review this properly but it would be great to also add a regression test.

thank u very much

File: app/backend/src/couchers/errors.py
Comment (kalvdans) [inline app/backend/src/couchers/errors.py]: Please follow the coding style and use spaces around `=`.

The `black` auto-formatter will do this for you.

File: app/backend/src/couchers/servicers/auth.py
Comment (kalvdans) [inline app/backend/src/couchers/servicers/auth.py]: Move these added lines two lines down so that the email check is not interrupted in the middle.

**Review by kalvdans — CHANGES_REQUESTED** [review]

**Comment by satvik6005** [comment]
@kalvdans i have made the changes now what

**Comment by kalvdans** [comment]
> @kalvdans i have made the changes now what

Great! Push your fixes to the pull request too.

**Review by satvik6005 — COMMENTED** [review]
fixed spaces and email is not anymore interrupted

**Comment by satvik6005** [comment]
> > @kalvdans i have made the changes now what
> 
> Great! Push your fixes to the pull request too.

i have done the work can u please review it

File: app/backend/src/couchers/servicers/auth.py
Comment (kalvdans) [inline app/backend/src/couchers/servicers/auth.py]: There's an extra space on the beginning of this line, which should have been discovered when running the unit tests or running black.

```
couchers/app/backend$ black src/couchers/servicers/auth.py
error: cannot format src/couchers/servicers/auth.py: Cannot parse: 165:0:                      existing_user_name = session.execute(
Oh no! 💥 💔 💥
1 file failed to reformat.
```

**Review by kalvdans — CHANGES_REQUESTED** [review]

File: app/backend/src/couchers/servicers/auth.py
Comment (satvik6005) [reply app/backend/src/couchers/servicers/auth.py]: done
Screenshot from 2021-08-19 20-01-33

Comment (satvik6005) [reply app/backend/src/couchers/servicers/auth.py]: i am sorry for all the delay this is my first time contributing to opensource

Comment (satvik6005) [reply app/backend/src/couchers/servicers/auth.py]: @kalvdans

Comment (kalvdans) [reply app/backend/src/couchers/servicers/auth.py]: No problem :) I'm just a bit curious since you clicked in the "All test pass" and "Formatted my code" in the checklist above.

However, my `black` still wants to fix more style in your file, are you sure you have pushed the latest version?
```diff
(venv) chn@brahe:~/couchers/couchers/app/backend$ black src/couchers/servicers/auth.py
reformatted src/couchers/servicers/auth.py
All done! ✨ 🍰 ✨
1 file reformatted.
(venv) chn@brahe:~/couchers/couchers/app/backend$ git diff
diff --git a/app/backend/src/couchers/servicers/auth.py b/app/backend/src/couchers/servicers/auth.py
index 21e25bad..1082fb65 100644
--- a/app/backend/src/couchers/servicers/auth.py
+++ b/app/backend/src/couchers/servicers/auth.py
@@ -154,20 +154,19 @@ class Auth(auth_pb2_grpc.AuthServicer):
                     existing_user = session.execute(
                         select(User).where(User.email == request.basic.email)
                     ).scalar_one_or_none()
-                   
 
                     if existing_user:
                         context.abort(grpc.StatusCode.FAILED_PRECONDITION, errors.SIGNUP_FLOW_EMAIL_TAKEN)
                     existing_flow = session.execute(
                         select(SignupFlow).where(SignupFlow.email == request.basic.email)
                     ).scalar_one_or_none()
-                    
+
                     existing_user_name = session.execute(
                         select(User).where(User.username == request.basic.username)
                     ).scalar_one_or_none()
                     if existing_user_name:
                         context.abort(grpc.StatusCode.FAILED_PRECONDITION, errors.SIGNUP_FLOW_USERNAME_TAKEN)
-                        
+
                     if existing_flow:
                         send_signup_email(existing_flow)
                         session.commit()
```

Comment (stifler6005) [reply app/backend/src/couchers/servicers/auth.py]: I only used black test. I will now use the test mentioned above...will that be all for formatting

Comment (satvik6005) [reply app/backend/src/couchers/servicers/auth.py]: Sorry i edited the soace...but forgot to push the file...will do asap @kalvdans

File: app/backend/src/couchers/servicers/auth.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/auth.py]: You've put the existing username check between the existing flow logic.

Move what you have to be above this `existing_flow = ...` statement

File: app/backend/src/couchers/servicers/auth.py
Comment (aapeliv) [inline app/backend/src/couchers/servicers/auth.py]: Should be `existing_username` with `username` as one word for consistency!

**Review by aapeliv — CHANGES_REQUESTED** [review]

File: app/backend/src/couchers/servicers/auth.py
Comment (satvik6005) [reply app/backend/src/couchers/servicers/auth.py]: i have moved above the existing flow logic

File: app/backend/src/couchers/servicers/auth.py
Comment (satvik6005) [reply app/backend/src/couchers/servicers/auth.py]: done

**Comment by satvik6005** [comment]
@kalvdans  can u please review the changes I made

**Comment by lucaslcode** [comment]
Hi @stifler6005, everyone here are volunteers so please allow some time for reviewing as people may not be available immediately. If you want to get going with something else in the meantime, your work here is saved so you can start work on another issue in a separate branch off `develop`. Thanks!

**Comment by satvik6005** [comment]
> Hi @stifler6005, everyone here are volunteers so please allow some time for reviewing as people may not be available immediately. If you want to get going with something else in the meantime, your work here is saved so you can start work on another issue in a separate branch off `develop`. Thanks!

ok...thnku for the feedback

**Review by kalvdans — APPROVED** [review]
Looks good to me! Thanks for your first contribution :)

There's a "Linked issues" list to the left in github. Please enter 1837 there and then the bug will automatically be marked fixed when you push the merge button.

**Comment by satvik6005** [comment]
> Looks good to me! Thanks for your first contribution :)
> 
> There's a "Linked issues" list to the left in github. Please enter 1837 there and then the bug will automatically be marked fixed when you push the merge button.

thank you for your approval appreciate it :)
there is "linked issue" written but it is not a link or button

**Comment by kalvdans** [comment]
> there is "linked issue" written but it is not a link or button

There's a cogwheel you can push to edit.

**Comment by satvik6005** [comment]
> > there is "linked issue" written but it is not a link or button
> 
> There's a cogwheel you can push to edit.

i am sorry but I cannot see it could u please send  me a screen short

**Comment by kalvdans** [comment]
> i am sorry but I cannot see it could u please send me a screen short

bild

**Comment by satvik6005** [comment]
> > i am sorry but I cannot see it could u please send me a screen short
> 
> bild

I don't think that I have the rights
Screenshot from 2021-08-20 15-23-17

**Comment by kalvdans** [comment]
> I don't think that I have the rights

True, I learn something new every day... 

Then I think it is just a matter of writing this text in a comment:
Closes #1837

(Seems it has to be the first comment, so I edited your first comment)

**Comment by satvik6005** [comment]
> > I don't think that I have the rights
> 
> True, I learn something new every day...
> 
> Then I think it is just a matter of writing this text in a comment:
> Closes #1837
> 
> (Seems it has to be the first comment, so I edited your first comment)

I just want to say....thank you:):)
now we have to wait for @aapeliv   fro approval

**Review by satvik6005 — COMMENTED** [review]
LGTM!

**Comment by aapeliv** [comment]
Reverted, see https://github.com/Couchers-org/couchers/issues/1837#issuecomment-902802148

**Pull request merged**
